src/GetDataStoreCapableTrait.php

```diff
@@ -23,12 +25,12 @@
      *
      * @since [*next-version*]
      *
-     * @return object The data store.
+     * @return ArrayObject The data store.
      */
     protected function _getDataStore()
     {
         return $this->dataStore === null
-                ? $this->dataStore = new \stdClass()
+                ? $this->dataStore = new ArrayObject()
```

Comment:
Why not abstract this into a `_createDataStore() : ArrayAccess` method? It would certainly be a beneficial abstraction to allow any `ArrayAccess` implementation. A `CreateDataStoreCapableTrait` can complement this.

Reply:
If you agree, be sure to change the `ArrayObject` types in the docs to `ArrayAccess`.

Reply:
Perhaps the abstraction could be useful. But let's talk about the hinted type.
For the implementations in this package, `ArrayAccess` is all that's necessary. However, a part of the reason for this whole changeset is the `MapInterface` implementation (and its countable version). The first requires the storage to in addition be `iterable`, and the second - `Countable`. What do we type-hint and where?

Reply:
I would still type hint `ArrayAccess` in this package. Like you said, it's all that's necessary here.

I anticipate that the map abstract code (be it a class or trait) will depend on some data store via an abstract method. _That_ method can document that it requires an iterable. The same applies for whatever abstract code will do the counting.

When the concrete or base classes are being composed, it is then the developer's job to make sure that everything is compatible. As long as the necessary abstraction is provided. A few checks also wouldn't hurt.

```php
class CountingAndStuff implements Countable
{
	// This class is now completely aware that the data store is an `ArrayObject`
	use CreateDataStoreCapableArrayObjectTrait;

    public function count()
    {
        $dataStore = $this->_getDataStore();

		// Still doesn't hurt to check, for completeness' sake. I actually encourage this anyway
		if ($dataStore instanceof Countable) {
			return $dataStore->count();
		}

		// Even if this might never happen, it's a good safeguard against potential instance setup problems
		throw new InvalidArgumentException('Data store is not countable.');
    }
}
```

Sure, that might not look anything like the final code, but it helps get my point across. Everything remains concerned with its own things. The alternative is to create a `DataStoreInterface`. But even then we'd probably end up segregating the interfaces and end up with the same problem and same solution.